### PR TITLE
Scoped Liquid TemplateContext

### PR DESCRIPTION
**Pull request by jtkech**
So, here the idea is to **create a shared liquid `TemplateContext` only once per scope** for perf, on wich we do once a shared contextualization and then a specific contextualization before each rendering.

- **Shared contextualization done once per scope**: Add scoped services in ambient values, call liquid handlers to add more ambient / scope values and access strategies, add all scoped liquid filters.

- **Specific contextualization before each rendering**: E.g contextualize the localizer with the current view context, update the specific model value and its access strategy.

Note: There are many places where we add a `ContentItem` value to the liquid scope, maybe we could remove them because now most of the time it is already accessible through `Model.ContentItem`.

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.Markdown/GraphQL/MarkdownBodyQueryObjectType.cs [hunk 1]
```diff
@@ -34,8 +34,8 @@ private static async Task<object> ToHtml(ResolveFieldContext<MarkdownBodyPart> c
             var htmlEncoder = context.ServiceProvider.GetService<HtmlEncoder>();
 
             var markdown = ctx.Source.Markdown;
-            var templateContext = new TemplateContext();
-            markdown = await liquidTemplateManager.RenderAsync(markdown, htmlEncoder, templateContext);
+            var templateContext = liquidTemplateManager.Context;
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Markdown/GraphQL/MarkdownBodyQueryObjectType.cs]: not necessary?

Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.Markdown/GraphQL/MarkdownBodyQueryObjectType.cs]: Oups, i will remove it.

**Comment by jtkech** [comment]
Just for infos

- In some `XxxPart[Field]QueryObjectType` we use the `liquidTemplateManager` but we can't pass a complete model because we work with a `ResolveFieldContext<XxxPart/Field>` that doesn't contain all properties, we would need a `BuildPart/FieldDisplayContext` as we pass to the related drivers

- E.g in `MarkdownFieldQueryObjectType` the `ResolveFieldContext<MarkdownField>` doesn't contain the `Part` to which the field belongs, we would need to work with a `BuildFieldDisplayContext`.

We will see if it is an issue.

**Comment by jtkech** [comment]
@sebastienros i think i'm ready with this one.

- I did some benchmark tests, seems to improve a little the rps, maybe around 5% for the blog page, but not sure because the results are not so stable on my machine.

- I will see in another PR what can be done for the liquid mvc bridge. Here i already did a 1st change in `HelperTag.cs` by checking `Statements.Count > 0` before building the child content that uses a `StringBuilderPool` and a `StringWriter`, knowing that most of the time tags have no child content.

**Comment by jtkech** [comment]
@sebastienros 

Based on how aspnetcore activate some properties by reflection, i tried a `fastPropertySetter` and compared it from our `setterDelegate`.

In a 100000 times loop | `setterDelegate` => 99 ms | `fastPropertySetter` => 3 ms

FastSetter

**Comment by jtkech** [comment]
For infos some average results on my machine on the blog home page.

Before

    .\ab -n 2000 -c 20  http://localhost:5000/ => around 1050 rps

After

    .\ab -n 2000 -c 20  http://localhost:5000/ => around 1150 rps => +10%

That said, for infos if i disable the layer feature

    .\ab -n 2000 -c 20  http://localhost:5000/ => around 1250 rps => +10% again

I will look into the layer feature.

**Comment by jtkech** [comment]
@sebastienros 

When you will have time, can you benchmark this one ;)

**Comment by jtkech** [comment]
@sebastienros for infos

- About `DynamicShapeBinding` i found the underlying issue that in fact was already fixed here after some refactoring. So, i could get rid of `DynamicShapeBinding` and we don't need anymore the code related to a special case and that was already there before. Now `LiquidViewTemplate.cs` has 100 lines less than before and `LiquidPage.cs` now is just one line of code.

- About the nested scope problem, before rendering we now use `EnterChildScope()` and then `ReleaseScope()`. Hmm, maybe the same problem with the `IViewLocalizer` that would be the last thing to check, maybe we would need to set it in the child scope, not in an ambient value.

**Comment by sebastienros** [comment]
Why did you remove the code with IndexOf, it was not working without it.

**Comment by jtkech** [comment]
But it works if, after having found a binding from the shape table, we don't check at all if a descriptor exists, because it exists necessarily, i left you a french comment on skipe.

Alteration buiders have a bindingName = shapeType with `__` (used to build a concrete shape) and a base shapeType without `__`. They are grouped by their base shapeType, and a given group alters the descriptor with this same base shapeType. So a descriptor of a given base shapeType provides multiple bindings whose bindingName may have `__` but always starts with this base shapeType without `__`.

In the **ShapeTable** we expose bindings directly to not have to lookup on descriptors **but the bindings collection is created from these descriptors**. So if a binding exists there is always a related descriptor, and the bindingName always starts with the base shapeType of this descriptor.

Note: bindings coming from shape resolvers doesn't have any related descriptors.

File: src/OrchardCore/OrchardCore.DisplayManagement.Liquid/HtmlContentWriter.cs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    public class HtmlContentWriter : TextWriter, IHtmlContent
+    {
+        private static readonly string[] _internedChars;
+        private const int _internedCharsLength = 256;
+
+        private readonly List<object> _fragments = new List<object>();
+
+        static HtmlContentWriter()
+        {
+            // Memorize all ASCII chars to prevent allocations
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.DisplayManagement.Liquid/HtmlContentWriter.cs]: https://en.wikipedia.org/wiki/Memoization

Comment (jtkech) [reply src/OrchardCore/OrchardCore.DisplayManagement.Liquid/HtmlContentWriter.cs]: Oups, updated

**Comment by stantoxt** [comment]
@jtkech I have the problem like this ,I don't know if it is about this feature
```
NullReferenceException: Object reference not set to an instance of an object.
Fluid.TemplateContext.ReleaseScope()
OrchardCore.DisplayManagement.Liquid.LiquidViewTemplateExtensions.RenderAsync(LiquidViewTemplate template, TextWriter writer, TextEncoder encoder, LiquidTemplateContext context, object model, Action<Scope> scopeAction) in LiquidViewTemplate.cs
-
            {
                await context.EnterScopeAsync(viewContext, model, scopeAction);
                await template.RenderAsync(writer, encoder, context);
            }
            finally
            {
                context.ReleaseScope();
            }
        }
        public static async Task<ViewContext> GetViewContextAsync(LiquidTemplateContext context)
        {
            var actionContext = context.Services.GetService<IActionContextAccessor>()?.ActionContext;
OrchardCore.Templates.Services.TemplatesShapeBindingResolver+<>c__DisplayClass8_0+<<BuildShapeBinding>b__0>d.MoveNext() in TemplatesShapeBindingResolver.cs
-
            {
                BindingName = shapeType,
                BindingSource = shapeType,
                BindingAsync = async displayContext =>
                {
                    var content = new ViewBufferTextWriterContent();
                    await _liquidTemplateManager.RenderAsync(template.Content, content, _htmlEncoder, displayContext.Value);
                    return content;
                }
            };
        }
    }
}
OrchardCore.DisplayManagement.Implementation.DefaultHtmlDisplay.<ProcessAsync>g__Awaited|12_0(Task<IHtmlContent> task) in DefaultHtmlDisplay.cs
-
        }
        private static ValueTask<IHtmlContent> ProcessAsync(ShapeBinding shapeBinding, IShape shape, DisplayContext context)
        {
            async ValueTask<IHtmlContent> Awaited(Task<IHtmlContent> task)
            {
                return CoerceHtmlString(await task);
            }
            if (shapeBinding?.BindingAsync == null)
            {
                // todo: create result from all child shapes
                return new ValueTask<IHtmlContent>(shape.Metadata.ChildContent ?? HtmlString.Empty);
```

**Pull request merged**